File: leetcode-04/solutions/python3/0768.py
class Solution:
    def maxChunksToSorted(self, arr: List[int]) -> int:
        n = len(arr)

        ans = 0
        min_ = [0] * (n - 1) + [arr[-1]]
        max_ = float('-inf')

        for i in range(n - 2, -1, -1):
            min_[i] = min(min_[i + 1], arr[i])

        for i in range(n - 1):
            max_ = max(max_, arr[i])
            if max_ <= min_[i + 1]:
                ans += 1

        return ans + 1

File: leetcode-04/solutions/cpp/0768.cpp
class Solution {
 public:
  int maxChunksToSorted(vector<int>& arr) {
    const int n = arr.size();

    int ans = 0;
    vector<int> min(n);
    min[n - 1] = arr[n - 1];
    int max = INT_MIN;

    for (int i = n - 2; i >= 0; i--) min[i] = std::min(min[i + 1], arr[i]);

    for (int i = 0; i < n - 1; i++) {
      max = std::max(max, arr[i]);
      if (max <= min[i + 1]) ans++;
    }

    return ans + 1;
  }
};